src/main/java/org/apache/hadoop/mapred/MesosScheduler.java

```diff
@@ -260,19 +252,28 @@ public synchronized void error(SchedulerDriver schedulerDriver, String s) {
     if (tasks != null) {
       HttpHost tracker = new HttpHost(taskTracker.getStatus().getHost(), taskTracker.getStatus().getHttpPort());
       MesosTracker mesosTracker = mesosTrackers.get(tracker);
-      synchronized (this) {
-        if (mesosTracker != null) {
-          if (mesosTracker.stopped) {
-            // Make sure we're not asked to assign tasks to any task trackers that have
-            // been stopped. This could happen while the task tracker has not been
-            // removed from the cluster e.g still in the heartbeat timeout period.
-            LOG.info("Asked to assign tasks to stopped tracker " + tracker + ".");
-            return null;
-          } else {
-            // Keep track of which TaskTracker contains which tasks.
-            for (Task task : tasks) {
-              mesosTracker.jobs.add(task.getJobID());
+      if (mesosTracker != null) {
+        synchronized (this) {
+          for (Iterator<Task> iterator = tasks.iterator(); iterator.hasNext();) {
+
+            // Throw away any task types that don't match up with the current known
+            // slot allocation of the tracker. We do this here because when we change
+            // the slot allocation of a running Task Tracker it can take time for
+            // this information to propagate around the system and we can preemptively
+            // avoid scheduling tasks to task trackers we know not to have capacity.
+            Task task = iterator.next();
+            if (task instanceof MapTask && mesosTracker.mapSlots == 0) {
```

Comment:
In this case, no, not that i'm aware of. Here we're only checking the value of the slots we've assigned to the `MesosTracker`, this value is always set to a positive int while processing offers and never changed again, except when set to zero. Perhaps `<=` would be safer, though.